src/main/scala/com/box/castle/batch/CastleMessageBatch.scala

```diff
@@ -74,6 +74,30 @@ case class CastleMessageBatch(messageAndOffsetSeq: IndexedSeq[MessageAndOffset],
     }
   }
 
+  /**
+    * Create a new CastleMessageBatch sliced off at a specific Size limit starting from the beginning.
+    * @param batchSize
+    * @return
+    */
+  def createBatchBySize(batchSize: Int): Option[CastleMessageBatch] = {
+    if(batchSize <= 0) None
+    else if (batchSize >= sizeInBytes) Some(this)
+    else {
+      val (endIdx, _) = messageAndOffsetSeq.foldLeft((0, batchSize)){ case ((idx, bSize), message) =>
+        val remainingSize = bSize - MessageSet.entrySize(message.message)
+        (if( remainingSize < 0 ) idx  else idx + 1, remainingSize)
```

Comment:
when remainingSize becomes less than zero, foldLeft will keep folding until end of the collection even though idx will not change, right?

Reply:
Yes.